mim/commands/install.py

```diff
@@ -1,580 +1,303 @@
+    find_link = WHEEL_URL['mmcv-full'].format(
+        cuda_version=cuda_v, torch_version=f'torch{torch_v}')
+    return find_link
 
-def infer_find_url(package: str) -> str:
-    """Try to infer find_url if possible.
 
-    If package is the official package, the find_url can be inferred.
+@contextmanager
+def patch_pkg_resources_distribution(
+        index_url: Optional[str] = None) -> Generator:
+    """A patch for `pip._vendor.pkg_resources.Distribution`.
+
+    Since the old version of the OpenMMLab package did not add the 'mim' extra
+    requirements to the release distribution, we need to hack the Distribution
+    and manually fetch the 'mim' requirements from `mminstall.txt`.
+
+    This patch works with 'pip<22.1' and 'pip>=22.1,python<3.11'.
 
     Args:
-        package (str): The name of package, such as mmcls.
+        index_url (str, optional): The pypi index url that pass to
+            `get_mmdeps_from_mmpkg_pypi`.
     """
-    find_url = ''
-    if package in WHEEL_URL:
-        torch_v, cuda_v = get_torch_cuda_version()
+    from pip._vendor.pkg_resources import Distribution, parse_requirements
+
+    origin_requires = Distribution.requires
 
-        # In order to avoid builiding mmcv-full from source, we ignore the
-        # difference among micro version because there are usually no big
-        # changes among micro version. For example, the mmcv-full built in
-        # pytorch 1.8.0 also works on 1.8.1 or other versions.
-        major, minor, *_ = torch_v.split('.')
-        torch_v = '.'.join([major, minor, '0'])
+    def patched_requires(self, extras=()):
+        deps = origin_requires(self, extras)
+        if self.project_name not in PKG2PROJECT or self.project_name == 'mmcv-full':  # noqa: E501
+            return deps
 
-        if cuda_v.isdigit():
-            cuda_v = f'cu{cuda_v}'
-        find_url = WHEEL_URL[package].format(
-            cuda_version=cuda_v, torch_version=f'torch{torch_v}')
-    elif package in PKG2PROJECT:
-        find_url = (f'{DEFAULT_URL}/{PKG2PROJECT[package]}.git')
+        if 'mim' in self.extras:
+            mim_extra_requires = origin_requires(self, ('mim', ))
+            filter_invalid_marker(mim_extra_requires)
+            deps += mim_extra_requires
+        else:
+            if not hasattr(self, '_mm_deps'):
+                assert self.version is not None
+                mmpkg = f'{self.project_name}=={self.version}'
+                mmdeps_text = get_mmdeps_from_mmpkg(mmpkg, index_url)
+                self._mm_deps = list(parse_requirements(mmdeps_text))
+                echo_warning(
+                    "Get 'mim' extra requirements from `mminstall.txt` "
+                    f'for {self}: {[str(i) for i in self._mm_deps]}.')
+            deps += self._mm_deps
+        return deps
 
-    return find_url
+    Distribution.requires = patched_requires  # type: ignore
+    yield
 
+    Distribution.requires = origin_requires  # type: ignore
+    return
```

Comment:
```suggestion
    Distribution.requires = patched_requires  # type: ignore
    try:
        yield
    finally:
        Distribution.requires = origin_requires  # type: ignore
```

Reply:
More details about why it is recommended to use `try...finally` can be found at https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager.

Reply:
Done!